### PR TITLE
[in-process] In-process linkage.

### PR DESCRIPTION
**Pull request by emankov**
+ Switched on by default.
+ In-process compilation is switched off for now (due to https://github.com/RadeonOpenCompute/ROCm-OpenCL-Driver/issues/13).
+ Temporary env var is added (till in-process compilation is switched off):

  export AMD_OCL_LINK_IN_PROCESS=0 to switch off.
  export AMD_OCL_LINK_IN_PROCESS=1 or unset AMD_OCL_LINK_IN_PROCESS to switch back on.

+ Testing done:
  ocl_conformance 1.2: api, atomics, basic, buffers, commonfns, compiler

### CODE REVIEW

File: src/driver/AmdCompiler.cpp [hunk 1]
```diff
@@ -79,6 +81,9 @@ using namespace llvm::object;
 using namespace clang;
 using namespace clang::driver;
 
+static cl::opt<bool>
+Verbose("v", cl::ZeroOrMore, cl::desc("Print information about actions taken"), cl::init(false));
```
Comment (rampitec) [inline src/driver/AmdCompiler.cpp]: The name of the option shall be changed. All options from linked in llvm have the same namespace, so -v is too generic.

File: src/driver/AmdCompiler.cpp [hunk 2]
```diff
@@ -354,7 +374,8 @@ AMDGPUCompiler::AMDGPUCompiler(const std::string& llvmBin_)
     llvmLinkExe(llvmBin + "/llvm-link"),
```
Comment (rampitec) [inline src/driver/AmdCompiler.cpp]: Can we perform in-process link without existing llvm bin dir?

File: src/driver/AmdCompiler.cpp [hunk 3]
```diff
@@ -655,14 +676,61 @@ bool AMDGPUCompiler::LinkLLVMBitcode(const std::vector<Data*>& inputs, Data* out
...
+        return false;
+      }
+      if (verifyModule(*m, &errs())) {
+        outs() << "ERROR: loaded module '" << arg << "' to link is broken!\n";
+        return false;
+      }
+      if (Verbose) {
+        std::cout << "Linking in '" << arg << "'\n";
+      }
+      if (L.linkInModule(std::move(m), ApplicableFlags)) {
```
Comment (rampitec) [inline src/driver/AmdCompiler.cpp]: Did you solve the problem you had with huge linked files with unreferenced symbols?

File: src/driver/AmdCompiler.cpp (hunk 2 above)
Comment (emankov) [inline src/driver/AmdCompiler.cpp]: Yes we can, llvLinkExe is used for Driver API path only. AMDGPUCompiler class will be refactored after finishing in-process.

File: src/driver/AmdCompiler.cpp (hunk 3 above)
Comment (emankov) [inline src/driver/AmdCompiler.cpp]: Yes I did, by using getLazyIRFileModule instead of parseIRFile.

File: src/driver/AmdCompiler.cpp (hunk 1 above)
Comment (emankov) [inline src/driver/AmdCompiler.cpp]: -v is option used by clang, llvm-link and others. The idea was to give a way to support the existing LLVM/Clang options in RocM OpenCL driver.

Comment (rampitec) [inline src/driver/AmdCompiler.cpp]: OK, they seem to be defined in main, so do not contradict.

File: src/driver/AmdCompiler.cpp (hunk 3 above)
Comment (rampitec) [inline src/driver/AmdCompiler.cpp]: Thanks. Good to know.

**Pull request merged**